src/Conversion/ONNXToKrnl/Math/MatMulInteger.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===--------- MatMulInteger.cpp - Lowering MatMulInteger Op --------------===//
+//
+// Copyright 2019-2023 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX MatMulInteger Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/DialectBuilder.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/Mlir/DialectBuilder.hpp"
+#include "src/Dialect/Mlir/IndexExpr.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXMatMulIntegerOpLowering
+    : public OpConversionPattern<ONNXMatMulIntegerOp> {
+public:
+  ONNXMatMulIntegerOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : OpConversionPattern(typeConverter, ctx) {}
+
+  LogicalResult matchAndRewrite(ONNXMatMulIntegerOp mmiOp,
+      ONNXMatMulIntegerOpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    using LocalDialectBuilder = MultiDialectBuilder<KrnlBuilder,
+        IndexExprBuilderForKrnl, MathBuilder, MemRefBuilder, OnnxBuilder>;
+    Operation *op = mmiOp.getOperation();
+    Location loc = ONNXLoc<ONNXMatMulIntegerOp>(op);
+    LocalDialectBuilder create(rewriter, loc);
+
+    ValueRange operands = adaptor.getOperands();
+    Value A = adaptor.getA();
+    Value B = adaptor.getB();
+    Value aZeroPoint = mmiOp.getAZeroPoint(); // Optional input.
+    Value bZeroPoint = mmiOp.getBZeroPoint(); // Optional input.
+
+    // Common types.
+    auto aType = A.getType().dyn_cast<ShapedType>();
+    auto aZeroPointType = aZeroPoint.getType().dyn_cast<ShapedType>();
+    auto resMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(mmiOp.getResult().getType()));
+    Type resElementType = resMemRefType.getElementType();
+
+    // Get shape.
+    ONNXMatMulIntegerOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    // Prepare input A.
+    Value AInt32 = create.onnx.cast(A, resElementType);
+    if (!isNoneValue(aZeroPoint)) {
+      Value aZeroPointInt32 = create.onnx.cast(aZeroPoint, resElementType);
+      int64_t aRank = aType.getRank();
+      int64_t aZeroPointRank = aZeroPointType.getRank();
+      // If broadcasting, e.g. A is [MxK], zeroPoint is [M], M != 1.
+      // Unsqueeze zeroPoint to [Mx1] to make shapes compatible.
+      // There is no need to handle scalar zeroPoint (e.g. tensor<dtype> or
+      // tensor<1xdtype>), which is always true for broadcasting.
+      if (((aZeroPointRank == 1) && (aZeroPointType.getShape()[0] != 1)) ||
+          ((aZeroPointRank > 1) && (aZeroPointRank = aRank - 1))) {
```

Comment:
given that verify() checks that aZeroPointRank == rank when aZeroPointRank > 1, I think you should only unsqueeze when aZeroPointRank == 1

otherwise the logic here disagrees with constant propagation which is inconsistent

Reply:
if there is only one element in the tensor (`aZeroPointRank == 1` and `aZeroPointShape[0] == 1`), we don't need to unsqueeze too. It's similar to scalar tensor.